### PR TITLE
[move] Use clever error locations for macro-rewritten lines in Move unit test errors

### PR DESCRIPTION
**Pull request by tzakian**
## Description 

This changes the way we report source locations for clever errors coming from an `assert!(<bool>)` within a macro. In particular, since the line number is rewritten during macro-expansion, we use this as the location of the error in unit tests when reporting the error if there is no overlap between the line number from the clever assertion error and the "actual" error.

Note that it was easiest to do this by plumbing in the `MappedFiles` instead of using the `FilesSourcesText` in unit tests anymore. This simplified a number of other places as well (e.g., for getting the line number from a location).

## Test plan 

Made sure existing tests continued to pass, and added tests for the new behavior.

---

## Release notes

Check each box that your changes affect. If none of the boxes relate to your changes, release notes aren't required.

For each box you select, include information after the relevant heading that describes the impact of your changes that a user might notice and any actions they must take to implement updates. 

- [ ] Protocol: 
- [ ] Nodes (Validators and Full nodes): 
- [ ] Indexer: 
- [ ] JSON-RPC: 
- [ ] GraphQL: 
- [ ] CLI: 
- [ ] Rust SDK:

### CODE REVIEW

File: external-crates/move/crates/move-cli/src/base/test.rs [hunk 1]
```diff
@@ -189,6 +172,7 @@ pub fn run_move_unit_tests<W: Write + Send>(
         let (mut compiler, cfgir) = compiler.into_ast();
         let compilation_env = compiler.compilation_env();
         let built_test_plan = construct_test_plan(compilation_env, Some(root_package), &cfgir);
+        let mapped_files = compilation_env.file_mapping().clone();
```
Comment (cgswords) [inline external-crates/move/crates/move-cli/src/base/test.rs:175]: :tada:

File: external-crates/move/crates/move-unit-test/src/test_reporter.rs [hunk 2]
```diff
@@ -61,7 +63,7 @@ pub struct TestStatistics {
     failed: BTreeMap<ModuleId, TestRuns<TestFailure>>,
 }
 
-#[derive(Debug, Clone)]
+// #[derive(Debug, Clone)]
```
Comment (cgswords) [inline external-crates/move/crates/move-unit-test/src/test_reporter.rs:66]: Why this change?

File: external-crates/move/crates/move-unit-test/src/test_reporter.rs [hunk 3]
```diff
@@ -110,6 +112,53 @@ impl FailureReason {
     }
 }
 
+fn clever_error_line_number_to_loc(test_plan: &TestPlan, vm_error: &VMError) -> Option<Loc> {
```
Comment (cgswords) [inline external-crates/move/crates/move-unit-test/src/test_reporter.rs:115]: Small nit,  but some of this functionality feels more general than the test reporter. Would it be worth migrating parts of it to `move-compiler/src/diagnostics/mod.rs`, possibly to live on the `PositionInfo` trait? I could see the IDE or similar benefitting from the ability to turn a line number into its definition loc , etc.

File: external-crates/move/crates/move-unit-test/src/test_reporter.rs [hunk 4]
```diff
@@ -110,6 +112,53 @@ impl FailureReason {
...
+            let mut start = line_range.start;
+            // Sub 1 to trim off newline character
+            let end = line_range.end - 1;
+
+            // Since we only have the line span, trim the start of the line to remove any
+            // whitespace.
+            let source_slice =
+                &test_plan.mapped_files.files().source(*file_id).unwrap()[start..end];
+            if let Some((whitespace, rest)) =
+                source_slice.split_once(|ch: char| !ch.is_ascii_whitespace())
```
Comment (cgswords) [inline external-crates/move/crates/move-unit-test/src/test_reporter.rs:148]: Would it be more efficient to call `trim_start()` and do the abs length comparison?

**Review by cgswords — APPROVED** [review]
LGTM modulo a few small nits. I'm glad the `MappedFiles` abstraction is getting more use.

File: external-crates/move/crates/move-unit-test/src/test_reporter.rs (hunk 2 above)
Comment (tzakian) [reply external-crates/move/crates/move-unit-test/src/test_reporter.rs:66]: Wasn't needed, so figured I'd remove it. Will remove the comment-out and just delete though.

File: external-crates/move/crates/move-unit-test/src/test_reporter.rs (hunk 3 above)
Comment (tzakian) [reply external-crates/move/crates/move-unit-test/src/test_reporter.rs:115]: Will do!

File: external-crates/move/crates/move-unit-test/src/test_reporter.rs (hunk 4 above)
Comment (tzakian) [reply external-crates/move/crates/move-unit-test/src/test_reporter.rs:148]: It would indeed...great idea!
